src/middlewared/middlewared/plugins/volume.py

```diff
@@ -0,0 +1,292 @@
+            break
+
+    return ret
+
+
+async def mount(dev, path, mntopts=None, fstype=None):
+    mount_cmd = ['/sbin/mount']
+
+    if isinstance(dev, str):
+        dev = dev.encode('utf-8')
+
+    if isinstance(path, str):
+        path = path.encode('utf-8')
+
+    if mntopts:
+        opts = ['-o', mntopts]
+    else:
+        opts = []
+
+    if fstype == 'ntfs':
+        mount_cmd = ['/usr/local/bin/ntfs-3g']
+        fstype = []
+    else:
+        fstype = ['-t', fstype] if fstype else []
+
+    proc = await Popen(
+        mount_cmd + opts + fstype + [dev, path],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        encoding='utf8',
+    )
+    output = await proc.communicate()
+
+    if proc.returncode != 0:
+        logger.debug("Mount failed (%s): %s", proc.returncode, output)
+        raise ValueError("Mount failed {0} -> {1}, {2}" .format(
+            proc.returncode,
+            output[0],
+            output[1]
+        ))
+    else:
+        return True
+
+
+async def umount(path, force=False):
+    if force:
+        cmdlst = ['/sbin/umount', '-f', path]
+    else:
+        cmdlst = ['/sbin/umount', path]
+
+    proc = await Popen(
+        cmdlst,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        encoding='utf8',
+    )
+    output = await proc.communicate()
+
+    if proc.returncode != 0:
+        logger.debug("Umount failed (%s): %s", proc.returncode, output)
+        raise ValueError("Unmount Failed {0} -> {1} {2}".format(
+            proc.returncode,
+            output[0],
+            output[1]
+        ))
+    else:
+        return True
+
+
+class KernelModuleContextManager:
+    def __init__(self, module):
+        self.module = module
+
+    async def __aenter__(self):
+        if self.module is not None:
+            if not await self.module_loaded():
+                await run('kldload', self.module, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+                if not await self.module_loaded():
+                    raise Exception('Kernel module %r failed to load', self.module)
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        if self.module is not None:
+            try:
+                await run('kldunload', self.module, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+            except Exception:
+                pass
+
+    async def module_loaded(self):
+        return (await run('kldstat', '-n', self.module, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=False)).returncode == 0
+
+
+class MountFsContextManager:
+    def __init__(self, dev, path, *args):
+        self.dev = dev
+        self.path = path
+        self.args = args
+
+    async def __aenter__(self):
+        await mount(self.dev, self.path, *self.args)
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        if await is_mounted(device=self.dev, path=self.path):
+            await umount(self.path)
+
+
+class JobProgressBuffer:
```

Comment:
This is a cool idea as well. Do you think we could leverage this somewhere else? Perhaps this could be moved to to job.py? Thoughts?